### PR TITLE
Adding Player Handles

### PR DESCRIPTION
**Pull request by faytey**
## Context (Problem, Motivation, Solution)

Link related issues!
Issue [117](https://github.com/0xFableOrg/0xFable/issues/117)

## Describe Your Changes
I added the smart contract to implement Player handles and wrote the tests to check the functions work properly

## Checklist

- [ ] I have performed a self-review of my code
- [ ] I ran `make check` and fixed resulting issues
- [ ] I ran the relevant tests and they all pass
- [ ] I wrote tests for my new features, or added regression tests for the bug I fixed

## Testing

If you didn't write tests, explain how you made sure the code was correct and working as intended.

### CODE REVIEW

**Comment by faytey** [comment]
I need your review on this before proceeding, I'm encountering errors setting up the web

**Comment by faytey** [comment]
> I need your review on this before proceeding @norswap , Also I'm encountering errors setting up the webapp

Hello @norswap still awaiting your review

File: package-lock.json
Comment (norswap) [inline package-lock.json:1]: Could you remove this file? We use `pnpm` for package management and we already have a pnpm-lock file in place!

File: packages/contracts/src/PlayerHandle.sol [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+        handles[msg.sender] = handle;
+        handleOwners[handle] = msg.sender;
+
+        emit HandleRegistered(msg.sender, handle);
+    }
+
+    function changeHandle(string memory newHandle) external {
+        require(checkHandleValidity(newHandle), "Invalid handle");
+        require(handleOwners[newHandle] == address(0), "Handle already taken");
```
Comment (norswap) [inline packages/contracts/src/PlayerHandle.sol:44]: Could you use custom errors for those?

File: packages/contracts/src/PlayerHandle.sol [hunk 2]
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {Ownable} from "openzeppelin/access/Ownable.sol";
+import {Strings} from "openzeppelin/utils/Strings.sol";
+
+interface IENSResolver {
+    function getEnsName(address addr) external view returns (string memory);
```
Comment (norswap) [inline packages/contracts/src/PlayerHandle.sol:8]: That doesn't seem to be the actual ENS resolver interface, which uses `name()` for lookups, see here: https://docs.ens.domains/resolvers/public

We also probably won't deploy on Ethereum, which is the only chain that has this resolver at the moment. ENS has announced that L2 support is coming but it's not there yet. I think it's fine to keep the resolver interface for now, at worst we can stub it with a permissioned registry that we sign on — let's just make the interface the same as the real thing.

File: packages/contracts/src/PlayerHandle.sol [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        }
+        return handles[player];
+    }
+
+    function checkHandleValidity(string memory handle) public pure returns (bool) {
+        bytes memory b = bytes(handle);
+        if (b.length < 5) return false;
+        for (uint256 i; i < b.length; i++) {
+            bytes1 char = b[i];
+            if (char < 0x20 || char > 0x7E || char == 0x2E) {
```
Comment (norswap) [inline packages/contracts/src/PlayerHandle.sol:74]: Note to myself: will probably want to change this to only support alphanumerics & dashes, underscores, square brackets (for cute clan names), spaces (but not more than one in a row, and not at the end or at the start of the handle), as well as put a max character limit.

Fine to leave this as is for this PR.

File: packages/contracts/src/test/PlayerHandle.t.sol [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    mapping(address => string) private ensNames;
+
+    function setEnsName(address addr, string memory name) external {
+        ensNames[addr] = name;
+    }
+
+    function getEnsName(address addr) external view override returns (string memory) {
+        return ensNames[addr];
+    }
+}
```
Comment (norswap) [inline packages/contracts/src/test/PlayerHandle.t.sol:17]: Could you spin this out in a separate file (in the test repo), and add a deployment of this + `PlayerHandle` to the `Deploy.s.sol`? You'll also want to add an entry for `PlayerHandle` in `deployment.ts` on the frontend.

**Review by norswap — CHANGES_REQUESTED** [review]
Great work! Sorry again for the delay. Could you make the requested changes? I'll make sure to prioritize reviewing them now.

**Comment by faytey** [comment]
Hi Sir I'd get started on these and update, sorry for the delay

**Comment by custosDiretriz** [comment]
Hi Sir, I've made the necessary corrections and pushed

**Comment by faytey** [comment]
Hi Sir, I have made requested corrections

**Comment by norswap** [comment]
Also it seems you've made changes to the lockfile, could just revert it to the master version so that no changes show? The current file seems incorrect in some way, as shown by the failing CI check.

**Comment by faytey** [comment]
> Also it seems you've made changes to the lockfile, could just revert it to the master version so that no changes show? The current file seems incorrect in some way, as shown by the failing CI check.

Hi Sir, I didn't make any changes to the lockfile and can't seem to understand how to rectify it.

**Comment by norswap** [comment]
That was probably accidental, you can see it is changed in "Files changes" here on Github.

Fixing it requires a little bit of git foo:
You can revert it by using `git checkout master -- pnpm-lock.yaml` then `git add pnpm-lock.yaml`.
You could then git commit and push that, but let's squash all your commit together:
`git reset --soft HEAD~7 && git commit`

(The 7 here is the number of commits you've made, so if you add additional commits in between you need to update that number.)

Then finally `git push --force-with-lease` to replace the commits here with the single squashed commit.

**Comment by faytey** [comment]
> That was probably accidental, you can see it is changed in "Files changes" here on Github.
> 
> Fixing it requires a little bit of git foo: You can revert it by using `git checkout master -- pnpm-lock.yaml` then `git add pnpm-lock.yaml`. You could then git commit and push that, but let's squash all your commit together: `git reset --soft HEAD~7 && git commit`
> 
> (The 7 here is the number of commits you've made, so if you add additional commits in between you need to update that number.)
> 
> Then finally `git push --force-with-lease` to replace the commits here with the single squashed commit.

done sir

**Comment by faytey** [comment]
Hello sir @norswap

**Comment by norswap** [comment]
Thanks! Did some extra cleanup and merged here: https://github.com/0xFableOrg/0xFable/pull/137

**Pull request closed without merging**